src/cargo/core/compiler/fingerprint.rs

```diff
@@ -249,64 +271,66 @@ pub fn prepare_target<'a, 'cfg>(
         source.verify(unit.pkg.package_id())?;
     }
 
-    let root = cx.files().out_dir(unit);
-    let missing_outputs = {
-        let t = FileTime::from_system_time(SystemTime::now());
-        if unit.mode.is_doc() {
-            !root
-                .join(unit.target.crate_name())
-                .join("index.html")
-                .exists()
-        } else {
-            match cx
-                .outputs(unit)?
-                .iter()
-                .filter(|output| output.flavor != FileFlavor::DebugInfo)
-                .find(|output| {
-                    if output.path.exists() {
-                        if mtime_on_use {
-                            // update the mtime so other cleaners know we used it
-                            let _ = filetime::set_file_times(&output.path, t, t);
-                        }
-                        false
-                    } else {
-                        true
-                    }
-                }) {
-                None => false,
-                Some(output) => {
-                    info!("missing output path {:?}", output.path);
-                    true
-                }
+    if compare.is_ok() && !force {
+        return Ok(Job::new(Work::noop(), Fresh));
+    }
+
+    let pkg_root = unit.pkg.root().to_path_buf();
+    let target_root = target_root(cx, unit);
+    let write_fingerprint = if unit.mode.is_run_custom_build() {
+        // For build scripts the `local` field of the fingerprint may change
+        // while we're executing it. For example it could be in the legacy
+        // "consider everything a dependency mode" and then we switch to "deps
+        // are explicitly specified" mode.
+        //
+        // To handlet his movement we need to regenerate the `local` field of a
+        // build script's fingerprint after it's executed. We do this by
+        // using the `build_script_local_fingerprints` function which returns a
+        // thunk we can invoke on a foreign thread to calculate this.
+        let state = Arc::clone(&cx.build_state);
+        let key = (unit.pkg.package_id(), unit.kind);
+        let (gen_local, _overridden) = build_script_local_fingerprints(cx, unit);
+        let output_path = cx.build_explicit_deps[unit].build_script_output.clone();
+        Work::new(move |_| {
+            let outputs = state.outputs.lock().unwrap();
+            let outputs = &outputs[&key];
+            let deps = BuildDeps::new(&output_path, Some(outputs));
+            if let Some(new_local) = gen_local.call_box(&deps, None)? {
+                // Note that the fingerprint status here is also somewhat
+                // tricky. We can't actually modify the `fingerprint`'s `local`
```

Comment:
I initially avoided that because I thought it would be too invasive, but after actually following through with it it actually feels quite good! I've switched.